### PR TITLE
[inductor] custom do_bench_gpu with smart cache flushing

### PR DESCRIPTION
**Pull request by nmacchioni**
TLDR; Triton do_bench blindly flushes 256MB to clear the L2 cache in between benchmarking iterations, which introduces a significant amount of compile time overhead for max-autotune mode. We can reduce this overhead by flushing a precise amount, equivalent to the current device's L2 cache size.

What we do in this PR:
- Add l2CacheSize to device properties
- Copy Triton's do_bench into Inductor

Key changes:
- Flushes the exact L2 cache size during benchmarking iterations. **NOTE**: We do not change the cache flushing in the estimation loop. The estimation loop includes the cache flushing overhead when calculating the number of warmup/benchmarking iterations; we do not change this behavior at all, so that the number of warmup/benchmarking iterations stays _exactly_ the same. 
- Remove unused do_bench_gpu parameters.

TODO:
- Remove cache flush overhead from estimation loop, and adjust (maybe cap) warmup/benchmark iterations appropriately. Subsequently, flush exact L2 cache size between estimation iterations.
- Remove fast_flush and default to always use fast flushing.
- Experiment with different quantile/return modes.

See also https://github.com/pytorch/pytorch/pull/127613
Base perf run https://github.com/pytorch/pytorch/actions/runs/9373796451
This perf run https://github.com/pytorch/pytorch/actions/runs/9375552247
cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC @MeetVadakkanchery @mhorowitz

### CODE REVIEW

File: torch/_inductor/runtime/runtime_utils.py [hunk 1]
```diff
@@ -81,35 +80,74 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
...
+
+    estimation_iters = 5
+    event_pairs = [
+        (
+            torch.cuda.Event(enable_timing=True),
+            torch.cuda.Event(enable_timing=True),
         )
+        for _ in range(estimation_iters)
+    ]
+    for start_event, end_event in event_pairs:
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:105]: nit: add comment that we are using the larger cache size to preserve BC wrt/number of benchmarking iterations for the triton do_bench function

File: torch/_inductor/runtime/runtime_utils.py [hunk 2]
```diff
@@ -81,35 +80,74 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
...
+        cache.zero_()
+        fn()
+        end_event.record()
+    torch.cuda.synchronize()
+    estimate_ms = min(
+        [event_pair[0].elapsed_time(event_pair[1]) for event_pair in event_pairs]
+    )
+
+    if fast_flush:
+        cache = torch.empty(int(get_cache_size() // 4), dtype=torch.int, device="cuda")
```
Comment (eellison) [inline torch/_inductor/runtime/runtime_utils.py:116]: nit: could we `del cache` above before reassigning ? it doesn't practically matter, but the cache size is large enough so that it interferes with our benchmarking memory compression ratios for torchbench models. 

double checked myself that they would both be live at the same time:
```
class Foo():
    def __init__(self):
        print("Init")

    def __del__(self):
        print("Del")

a = Foo()
a = Foo()
```

```
Init
Init
Del
Del
```

**Review by eellison — APPROVED** [review]
looks good !

File: torch/_inductor/runtime/runtime_utils.py (hunk 2 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/runtime_utils.py:116]: Sure; it would probably also help to manually delete the new cache at the end of benchmarking as well. I wonder if there is anything else like this messing with the memory compression ratios, might be worth looking at.
